47 standardize drift and stt outputs of digitization

valerpia opened the pull request.

valerpia commented on CMakeLists.txt:
```diff
@@ -7,17 +7,19 @@
 project(SandReco LANGUAGES CXX VERSION 01.00.00)
 
-set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD 14)
```
back to 17

valerpia commented on CMakeLists.txt:
```diff
@@ -61,9 +93,30 @@ target_link_libraries(Digitize Struct SANDGeoManager Utils)
+# # Creates ReconstructNLLmethod executable for testing.
+# add_executable(test_reconstructNLL tests/test_reconstructionNLL.cpp)
+# target_link_libraries(test_reconstructNLL Struct Utils SANDRecoUtils)
+# ----
```
Remove

valerpia commented on README.md:
````diff
@@ -144,6 +144,16 @@ $ SANDECALClustering -d <digit file>
+```console
+$ ./build/bin/ReconstructNLLmethod -edep <EDEP file> -wireinfo tests/wireinfo.txt -o <reco file> --hit_time --signal_propagation
+```
+
````
remove

valerpia commented on include/SANDEventDisplay.h:
```diff
@@ -119,6 +131,8 @@ class SANDEventDisplay : public TGMainFrame
   std::vector<EVHits_t> fEventHitsZX;
   std::vector<EVHits_t> fTubeDigitHitsZY;
   std::vector<EVHits_t> fTubeDigitHitsZX;
```
check if needed

valerpia replied:
Replaced with fWireDigitHitsZY/X

valerpia commented on include/SANDRecoUtilsLinkDef.h:
Check if needed

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -1,9 +1,12 @@
 #include "SANDECALCellInfo.h"
-#include "SANDSTTTubeInfo.h"
+#include "SANDWireInfo.h"
+#include "SANDTrackerModule.h"
```
This should be SANDTrackePlane.h

valerpia replied:
The SANDTrackerModuleID in this header file is needed in the SANDManager.
The SANDTrackerModule class is not but I'll keep it for possible future uses

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -12,28 +15,59 @@
-    // "STT_([0-9]+)_(Trk|C3H6|C)Mod_PV_([0-9]+)(/|)";
     "(C|C3H6|Trk)Mod_([0-9]+)_PV_([0-9]+)(/|)";
+
+// const char* const stt_single_tube_regex_string =
```
remove comments

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -75,24 +109,45 @@ const double endcap_thickness = 115.0;
+  std::map<std::string, int> hit_counter_;
+  void IncrementCounter(std::string k);
+  void PrintCounter();
+};
```
Remove

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -75,24 +109,45 @@ const double endcap_thickness = 115.0;
-  std::map<int, SANDSTTTubeInfo> sttmap_;    // map of stt tube (key: id, value:
-                                             // info on tube)
+
+  std::map<SANDWireID, SANDWireInfo> wiremap_;  // map of wire (key : id, value:
```
remove

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -75,24 +109,45 @@ const double endcap_thickness = 115.0;
+  std::map<SANDWireID, SANDWireInfo> wiremap_;  // map of wire (key : id, value:
+                                          // info on wire)
+
+  std::map<long, std::map<double, long>>
```
old? Remove

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -75,24 +109,45 @@ const double endcap_thickness = 115.0;
+                                                          // info about tube
+                                                          // from volume path
+  // mutable TPRegexp stt_two_tubes_regex_{
+  // sand_geometry::stt::stt_two_tubes_regex_string};  // regular expression to
```
remove comments

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -102,10 +157,29 @@ class SANDGeoManager : public TObject
+
+
+  bool getLineSegmentIntersection(TVector2 p, TVector2 dir, TVector2 A, TVector2 B, TVector3& intersection);
+  bool getLineSegmentIntersection(TVector2 p, TVector2 dir, TVector2 A, TVector2 B, TVector2& intersection);
```
Could these be in utlis?

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -136,27 +210,58 @@ class SANDGeoManager : public TObject
                                      int& cell_local_id) const;
   void set_ecal_info();
 
+  void set_wire_info();
```
check if needed

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -136,27 +210,58 @@ class SANDGeoManager : public TObject
+                                                       const SANDTrackerPlane& plane);
+  std::vector<TVector2> getGlobalLinePlaneIntersections(const TVector2& local_2d_position, 
+                                                        const SANDTrackerPlane& plane);
+  double getMinDistanceBetweenSegments(TVector3 a, TVector3 b,
```
utils?

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -136,27 +210,58 @@ class SANDGeoManager : public TObject
-        stt_tube_regex_(sand_geometry::stt::stt_tube_regex_string),
-        stt_two_tubes_regex_(sand_geometry::stt::stt_two_tubes_regex_string),
+        wiremap_(),
+        wire_tranverse_position_map_(),
```
check if neded

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -136,27 +210,58 @@ class SANDGeoManager : public TObject
+  void SetGeoCurrentDirection(double x, double y, double z) const;
+  void InitVolume(volume& v) const;
+  void LOGVolumeInfo(volume& v) const;
+  void PrintCounter();
```
class counter? Remove

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -136,27 +210,58 @@ class SANDGeoManager : public TObject
   {
   }
   void init(TGeoManager* const geo);
+  void SetGeoCurrentPoint(double x, double y, double z) const;
```
check if needed

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -166,37 +271,89 @@ class SANDGeoManager : public TObject
     return cellmap_;
   }
-  const std::map<int, SANDSTTTubeInfo>& get_stt_tube_info() const
+  const std::map<SANDWireID, SANDWireInfo>& get_wire_info() const
```
check if needed

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -166,37 +271,89 @@ class SANDGeoManager : public TObject
+  {
+    return wiremap_;
+  }
+  const std::map<long, std::map<double, long>>&
```
check if needed

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -166,37 +271,89 @@ class SANDGeoManager : public TObject
+  std::vector<SANDTrackerCellID> get_segment_ids(const TG4HitSegment& hseg) const;
+  TVector3 FindClosestDrift(TVector3 point, double epsilon) const;
+  TVector3 SmearPoint(TVector3 point, double epsilon) const;
+  bool IsOnEdge(TVector3 point) const;
```
check if needed

valerpia commented on include/SANDGeoManager.h:
```diff
@@ -166,37 +271,89 @@ class SANDGeoManager : public TObject
                                   int& module_id, int& layer_id,
                                   int& cell_local_id);
+  TString FindNextActiveLayer(const double* starting_point,
+                              const double* direction) const;
```
check if needed

valerpia commented on include/SANDTrackerCell.h:
```diff
@@ -0,0 +1,153 @@
+  double _width;
+  double _height;
+
+  double _timeResponse;   // time response in nanoseconds
```
check if needed

valerpia commented on include/SANDTrackerCell.h:
```diff
@@ -0,0 +1,153 @@
+  void isFired(bool fired)
+  {
+    _isFired = fired;
+  }
```
check if needed

valerpia commented on include/SANDTrackerCell.h:
```diff
@@ -0,0 +1,153 @@
+    _id = wID;
+  }
+
+  double timeResponse() const
```
check if needed

valerpia commented on include/SANDTrackerModule.h:
remove?

valerpia commented on include/SANDTrackerUtils.h:
```diff
@@ -0,0 +1,276 @@
+  ~SANDTrackerUtils(){};
+  static void Clear();
+  static bool AreAdjacent(const SANDTrackerCellID &tub1, const SANDTrackerCellID &tub2);
+  static const SANDTrackerCellID GetTubeID(const SANDTrackerDigitID &id);
```
check if needed

valerpia replied:
check naming

valerpia commented on include/SANDTrackerUtils.h:
```diff
@@ -0,0 +1,276 @@
+
+
+
+class TrajectoryParameters {
```
check if needed

valerpia replied:
See comment at the beginning of the file

valerpia commented on include/SANDTrackerUtils.h:
check what is not needed anymore

valerpia replied:
There are things that will likely not be used again but they were used to validate the KF algorithm. I'm keeping them until we complete the new validation

valerpia commented on include/SANDTrackletFinder.h:
```diff
@@ -0,0 +1,83 @@
+    void Draw3DWires();
+    void Draw2DWires();
+    void Draw2DDistance();
+    void Draw2DDigits();
```
check if needed

valerpia replied:
Not strictly needed but not feeling good removing them right now

valerpia commented on include/SANDWireInfo.h:
```diff
@@ -0,0 +1,134 @@
+  enum class Orient {
+    kHorizontal,
+    kVertical
+  };
```
remove?

valerpia commented on include/SANDWireInfo.h:
```diff
@@ -0,0 +1,134 @@
+  SANDWireID id_;                  // id of tube
+  double x_;                // x position of the center of the tube
+  double y_;                // y position of the center of the tube
+  double z_;                // z position of the center of the tube
```
check if needed

valerpia commented on include/SANDWireInfo.h:
```diff
@@ -0,0 +1,134 @@
+  double ax_;
+  double ay_;
+  double az_;
+  std::vector<TVector3> points;
```
this should be a pair (not std) of points or an array (std)

valerpia replied:
I still think this is true, but currently I don't have any good idea

valerpia commented on include/struct.h:
```diff
@@ -93,11 +119,12 @@ struct cluster {
   double vary;
   double varz;
-  //std::vector<dg_cell> cells;
+  std::vector<dg_cell> cells;
```
remove

valerpia replied:
Can't remove it as it breaks analysis.cpp otherwise. Calo people are working to fix analysis in a different branch. Keeping it in the meanwhile

valerpia commented on include/struct.h:
```diff
@@ -170,12 +199,20 @@ struct event {
+  TGeoVolume* geo_volume;
+  TString volume_path;
+  bool IsActive;
+};
```
check if needed

valerpia replied:
used by calo people

valerpia commented on include/utils.h:
```diff
@@ -14,6 +14,7 @@
 
 namespace sand_reco
 {
+extern std::map<int, double> t0;
```
check if needed

valerpia replied:
it is used by FLUKA digitization. We need to update the digitization before removing this

valerpia commented on include/utils.h:
```diff
@@ -79,11 +79,16 @@ extern std::map<int, double> t0;
 
+namespace chamber
+{
+void initT0(TG4Event* ev, SANDGeoManager& geo);
```
check if needed

valerpia commented on src/SANDDigitization.cpp:
```diff
@@ -47,7 +47,7 @@ C                      + 1ns  uncertainty
     std::cout << "scint: " << tdec << std::endl;
     std::cout << "prop : "
-              << sand_reco::ecal::scintillation::vlfb * d * conversion::mm_to_m
+              << sand_reco::ecal::scintillation::vlfb* d* conversion::mm_to_m
```
fix spaces

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -25,7 +167,7 @@ double energy_to_photo_electrons(double E)
   if (debug)
     std::cout << "E = " << E
-              << " -> p.e. = " << sand_reco::ecal::photo_sensor::e2pe * E
+              << " -> p.e. = " << sand_reco::ecal::photo_sensor::e2pe* E
```
space

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -62,11 +204,33 @@ bool process_hit(const SANDGeoManager& g, const TG4HitSegment& hit, int& detID,
 
   g.decode_ecal_cell_id(cell_global_id, detID, modID, planeID, cellID);
+
+  if (cellID == -99) {
```
remove

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -201,11 +358,10 @@ void group_pmts_in_cells(const SANDGeoManager& geo,
        it != ps.end(); ++it) {
     int id = abs(it->first);
-
+    // if(id==999) continue;
```
remove

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -217,6 +373,7 @@ void group_pmts_in_cells(const SANDGeoManager& geo,
     } else {
       c->ps2 = it->second;
     }
+    if (c->id == 214294) continue;
```
what? why? how? then remove

valerpia replied:
magie di gianfranco

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -276,10 +433,12 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
+    SANDTrackerCellID stid = geo.get_stt_tube_id(x, y, z);
+
     if (stid == -999) {
       // std::cout << std::setprecision(12) << x << " " << y << " " << z <<
```
rmeove

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -313,49 +472,56 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
 // for each tube simulate tdc and adc
 // tdc is the time of closest point to wire + drift time
 // adc is the sum of energy deposit within integration time window
 void create_digits_from_hits(const SANDGeoManager& geo,
```
remove

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -392,38 +558,257 @@ void create_digits_from_hits(const SANDGeoManager& geo,
+  return ((point - middle).Mag() <= (start - middle).Mag());
+}
+
+
```
check if needed

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -506,17 +898,23 @@ void digitize(const char* finname, const char* foutname,
   std::cout << "\b\b\b\b\b" << std::setw(3) << 100 << "%]" << std::flush;
   std::cout << std::endl;
 
+  sand_geo.PrintCounter();
```
check if needed and remove

valerpia commented on src/SANDDigitizationEDEPSIM.cpp:
```diff
@@ -530,7 +928,7 @@ void digitize(const char* finname, const char* foutname,
   // sand_reco::stt::stPos.clear();
   // sand_reco::stt::tubePos.clear();
-  sand_reco::stt::t0.clear();
+  sand_reco::t0.clear();
```
check if needed and remove

valerpia replied:
calo people

valerpia commented on src/SANDDigitizeDrift.cpp:
remove

valerpia commented on src/SANDEventDisplay.cpp:
è rimasto li un po' così. check if working?

valerpia commented on src/SANDMeasurementsBuilder.cpp:
remove

valerpia replied:
this should be in reconstruct somehow

valerpia commented on src/SANDRecoUtils.cpp:
remove

valerpia commented on src/SANDTrackerModule.cpp:
rmeove?

valerpia commented on src/SANDTrackletFinder.cpp:
```diff
@@ -0,0 +1,385 @@
+  double y_width = _cells_intersections[1].Y() - _cells_intersections[0].Y();
+  
+  // To Do: should be a config parameter
+  int subdivisions = 3;
```
reduce to 2

valerpia commented on src/SANDTrackletFinder.cpp:
```diff
@@ -0,0 +1,385 @@
+  // std::cout << "SAMPLING SIZE: " << sampling_points.size() << std::endl;
+  for (uint i = 0; i < sampling_points.size(); i++) {
+    double starting_point[4] = {sampling_points[i][0], sampling_points[i][1], sampling_points[i][2], sampling_points[i][3]};
+    minimizer->SetLimitedVariable(0, "px", starting_point[0], 0.01, _cells_intersections[0].X() - 200, _cells_intersections[1].X() + 200);
```
reduce size of step

valerpia replied:
O(mm)
O(0.1 rad)

valerpia commented on src/display.cpp:
check if working

valerpia commented on src/reconstruction.cpp:
check if working

valerpia commented on src/reconstructionNLLmethod:
remove

valerpia commented on src/reconstructionNLLmethod.cpp:
rmeove

valerpia commented on src/utils.cpp:
```diff
@@ -792,14 +801,36 @@ void sand_reco::stt::initT0(TG4Event* ev, SANDGeoManager& geo)
+  }
+}
+
+void sand_reco::chamber::initT0(TG4Event* ev, SANDGeoManager& geo)
```
check if needed

valerpia replied:
they will at some point, but not like this. Removed

valerpia commented on tests/python_tools/EventDisplay.ipynb:
remove

valerpia commented on tests/test_DisplayReco.cpp:
remove

valerpia commented on wireinfo.txt:
remove

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -123,6 +137,38 @@ void SANDGeoManager::decode_ecal_cell_id(int cell_global_id, int& detector_id,
   cell_local_id = cell_global_id;
 }
 
+TString SANDGeoManager::FindNextActiveLayer(const double* starting_point,
```
check if needed

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
-void SANDGeoManager::set_stt_tube_info(const TGeoNode* const node,
-                                       const TGeoHMatrix& matrix,
-                                       int stt_plane_id)
+std::vector<TVector2> SANDGeoManager::getLocalLinePlaneIntersections(const TVector2& local_2d_position,
```
chenge name

valerpia replied:
Naa, I like it today

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
+}
+
+
+void SANDGeoManager::set_wire_info(const TGeoHMatrix& matrix)
```
setPlaneInfo()

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
+  file_wireinfo.close();
+}
+
+void Counter::IncrementCounter(std::string k)
```
rmeove

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
 {
   geo_ = geo;
-
+  counter_.hit_counter_.clear();
```
remove

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
-  sttmap_.clear();
-  stt_tube_tranverse_position_map_.clear();
-
+  wiremap_.clear();
```
rmeove

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
+  set_wire_info();
+}
+
+void SANDGeoManager::SetGeoCurrentPoint(double x, double y, double z) const
```
check if needed

valerpia replied:
used by calo people

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
+  }
+}
+
+void SANDGeoManager::LOGVolumeInfo(volume& v) const
```
fino a qui dal commento sopra

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -414,165 +531,842 @@ bool SANDGeoManager::is_stt_plane(const TString& volume_name) const
   TString volume_name = node->GetName();
   TString volume_path = geo_->GetPath();
 
+  if (!volume_name.Contains("Active")) {
```
cleanup

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -600,76 +1394,223 @@ int SANDGeoManager::get_ecal_cell_id(double x, double y, double z) const
 }
 
-int SANDGeoManager::get_stt_tube_id(double x, double y, double z) const
+double SANDGeoManager::GetHitCellDistance(TVector2 rotated_yz_hit_position, 
```
wire not cell

valerpia replied:
This is wrong. No changes needed

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -600,76 +1394,223 @@ int SANDGeoManager::get_ecal_cell_id(double x, double y, double z) const
+  return (rotated_yz_hit_position - rotated_yz_wire_position).Mod();
+}
+
+SANDTrackerCellID SANDGeoManager::GetClosestCellToHit(TVector3 hit_center, const SANDTrackerPlane& plane, bool checkCloseCells = false) const
```
re-undertand this piece of code

valerpia replied:
I remember this was due to some strange behaviours with the STT.
I haven't checked what the problem is so I'm adding a To Do comment here

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -600,76 +1394,223 @@ int SANDGeoManager::get_ecal_cell_id(double x, double y, double z) const
-    tube_id = stt_tube_tranverse_position_map_.at(plane_id).begin()->second;
-  } else if (it == stt_tube_tranverse_position_map_.at(plane_id).end()) {
-    tube_id = stt_tube_tranverse_position_map_.at(plane_id).rbegin()->second;
+long SANDGeoManager::get_wire_id(long drift_plane_id, double z,
```
re-uynderstand

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -600,76 +1394,223 @@ int SANDGeoManager::get_ecal_cell_id(double x, double y, double z) const
+  return wire_id;
+}
+
+bool SANDGeoManager::IsOnEdge(TVector3 point) const
```
check if needed then remove

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -600,76 +1394,223 @@ int SANDGeoManager::get_ecal_cell_id(double x, double y, double z) const
+  return OnEdge;
+}
+
+TVector3 SANDGeoManager::SmearPoint(TVector3 point, double epsilon) const
```
remove probably

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -600,76 +1394,223 @@ int SANDGeoManager::get_ecal_cell_id(double x, double y, double z) const
-}
\ No newline at end of file
+  // What are these?
+  IsOnEdge(hseg.Start.Vect());
```
remove

valerpia commented on src/SANDGeoManager.cpp:
```diff
@@ -600,76 +1394,223 @@ int SANDGeoManager::get_ecal_cell_id(double x, double y, double z) const
+
+  auto middle = (hseg.Start + hseg.Stop) * 0.5;
+
+  if (IsOnEdge(middle.Vect())) {
```
rmeove

valerpia reviewed (commented):
stt:numeric_limits<T>::max()

gabriele-sirri commented on include/SANDTrackerModuleConfig.h:
```diff
@@ -0,0 +1,57 @@
+
+namespace TrackerModuleConfiguration 
+{
+  namespace Drift
```
We should agree that namespace names are all lower-case.

Lowercase in namespaces are more popular and make it much easier to distinguish what’s a **namespace** and what’s a **Class**.

This convention is also suggested here:
https://google.github.io/styleguide/cppguide.html#Namespace_Names

gabriele-sirri commented on include/SANDTrackerModuleConfig.h:
```diff
@@ -0,0 +1,57 @@
+    };
+  }
+  
+  namespace STT
```
lowercase

gabriele-sirri commented on include/SANDTrackerPlane.h:
```diff
@@ -0,0 +1,105 @@
+#include "SANDTrackerCell.h"
+#include <vector>
+#include <map>
+class SANDTrackerModule;
```
Why not defining some namespaces to keep the Class name shorter, as for instance:

`sand::tracker::Module`

namespaces have pros and cons but brings benefit both to readibility and scoping, as suggested here: https://google.github.io/styleguide/cppguide.html#Namespaces

gabriele-sirri commented on include/SANDTrackerUtils.h:
```diff
@@ -0,0 +1,247 @@
+#include "SANDTrackerDigitCollection.h"
+#include "SANDKFTrack.h"
+
+using SANDTrackANDKFStateCovarianceMatrix = TMatrixD;
```
Would be useless if classes were defined in one (same) namespace

gabriele-sirri commented on include/SANDTrackerUtils.h:
```diff
@@ -0,0 +1,247 @@
+class SANDTrackerUtils
+{
+ private:
+  static TGeoManager* fGeo;
```
variable inside a class should follow a (the same) naming convention

please see: https://google.github.io/styleguide/cppguide.html#Variable_Names

gabriele-sirri commented on Makefile:
```diff
@@ -13,8 +13,8 @@ EDEPINCDIR = $(EDEPSIM)/include/EDepSim
 
-struct.cxx: include/struct.h include/Linkdef.h
-	cd include && rootcint -f ../src/$@ -c $(CFLAGS) -p $(HEADERS) Linkdef.h && cd ..
+struct.cxx: include/struct.h include/StructLinkdef.h
```
Usually, it doesn't make sense to have both a Makefile and a CMakeLists.txt in the same folder. Why is this Makefile here, and why is it being maintained?

gabriele-sirri commented on include/CLine3D.h:
```diff
@@ -0,0 +1,36 @@
+
+#include "TVector3.h"
+
+class CLine3D
```
I would improve readability by removing the 'C' prefix, as the coding convention proposed for sand-reco does not require it.

https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

gabriele-sirri commented on include/CLine3D.h:
```diff
@@ -0,0 +1,36 @@
+  TVector3 _point;
+  TVector3 _direction;
+  TVector3 _u;
+  TVector3 _v;
```
We need to standardize the variable names. For now, I would follow the style initially proposed in a minimal way, without introducing a new one for class data members.

https://google.github.io/styleguide/cppguide.html#Variable_Names

which suggests trailing underscores

gabriele-sirri commented on include/CLine3D.h:
```diff
@@ -0,0 +1,36 @@
+class CLine3D
+{
+ private:
+  TVector3 _point;
```
I'm skeptical about relying on external software, including ROOT. Is it truly necessary to use a type that inherits from TObject as a member of a class that is not inherited from TObject? or are we only utilizing a few features that we could implement ourselves?

In addition, TVector3 is not recommended by ROOT itself...(read "Attention" in https://root.cern.ch/doc/master/classTVector3.html)

gabriele-sirri commented on include/CLine3D.h:
```diff
@@ -0,0 +1,36 @@
+
+ public:
+  CLine3D() {};
+  CLine3D(const TVector3 &p, const TVector3 &d);
```
I know it's a matter of personal preference, but as a C++ programmer, I would emphasize types over syntax. 

`Line3D(const TVector3& p, const TVector3& d)`

Or even better, in my opinion, to make it more readable (to be read in reverse order as _p is a const reference to a TVector3_ ):

`Line3D(TVector3 const& p, TVector3 const& d)`

https://isocpp.org/wiki/faq/style-and-techniques#whitespace

gabriele-sirri commented on src/SANDEventDisplay.cpp:
```diff
@@ -30,37 +30,27 @@ using namespace std;
-                            "",
-                            "exit.xpm",
-                            0};
+enum ETestCommandIdentifiers {
```
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#enum5-dont-use-all_caps-for-enumerators

and here

https://google.github.io/styleguide/cppguide.html#Enumerator_Names

gabriele-sirri commented on include/SANDDigitizationEDEPSIM.h:
```diff
@@ -48,12 +48,34 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
+
+bool isInHit(hit& h, TVector3& point);
+
+std::vector<TLorentzVector> WireHitClosestPoints(hit& h,
```
It would also be necessary to establish a naming convention for function names. Personally, I would favor following Stroustrup's guidelines rather than the Google Style, especially considering the majority of the existing code.

Do not capitalize function names. Capitalize the names of your user-defined types. 
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-knr

gabriele-sirri commented on include/SANDDigitizationEDEPSIM.h:
```diff
@@ -48,12 +48,34 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
+void group_hits_by_wire(TG4Event* ev, const SANDGeoManager& geo,
+                        std::map<int, std::vector<hit> >& hits2wire);
+
+bool isInWire(SANDWireInfo& wire, TVector3& point);
```
From the function name, it seems that neither the wire nor the point will be modified.

Anything more complex than an `int `or a `double `should be passed as a `const&` if it is not an IN-OUT argument.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const

gabriele-sirri commented on include/SANDDigitizationEDEPSIM.h:
```diff
@@ -48,12 +48,34 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
 
+namespace chamber
+{
+void group_hits_by_wire(TG4Event* ev, const SANDGeoManager& geo,
```
Use `TG4Event&  ev`  or `TG4Event const&  ev` instead of `TG4Event* ev`
Passing an argument as a pointer makes sense if the object is expected to possibly be `nullptr `and the function includes a check for it. Otherwise,  it would be preferable to pass the object by reference.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f60-prefer-t-over-t-when-no-argument-is-a-valid-option

This comment applies to other parts of the code as well.

gabriele-sirri commented on include/SANDDigitizationEDEPSIM.h:
```diff
@@ -48,12 +48,34 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
+std::vector<TLorentzVector> WireHitClosestPoints(hit& h,
+                                                 SANDWireInfo& arg_wire);
+
+double GetMinWireTime(TLorentzVector point, SANDWireInfo& arg_wire);
```
`double get_min_wire_time(TLorentzVector const& point, SANDWireInfo const& arg_wire);`

gabriele-sirri commented on include/SANDDigitization.h:
```diff
@@ -10,8 +10,14 @@ using namespace sand_reco;
 
-enum class DETSIM_TYPE { kEdepsim, kFluka };
-enum class ECAL_digi_mode { const_fract, fixed_thresh };
+enum class DETSIM_TYPE {
```
ALL CAPS names should be used for MACROs.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es9-avoid-all_caps-names

gabriele-sirri commented on include/SANDGeoManager.h:
```diff
@@ -102,10 +119,28 @@ class SANDGeoManager : public TObject
+  void DrawModulesInfo();
+
+  // DRIFT CHAMBER
+  mutable TPRegexp wire_regex_{sand_geometry::chamber::wire_regex_string};
```
why `mutable `?

gabriele-sirri commented on include/SANDDigitizationEDEPSIM.h:
```diff
@@ -48,12 +48,34 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
+                                  std::map<int, std::vector<hit> >& hits2wire,
+                                  std::vector<dg_wire>& wire_digits);
+
+void digitize_wire(TG4Event* ev, const SANDGeoManager& geo,
```
`TG4Event const& ev`

gabriele-sirri commented on include/SANDDigitizationEDEPSIM.h:
```diff
@@ -48,12 +48,34 @@ void group_hits_by_tube(TG4Event* ev, const SANDGeoManager& geo,
+double GetMinWireTime(TLorentzVector point, SANDWireInfo& arg_wire);
+
+void create_digits_from_wire_hits(const SANDGeoManager& geo,
+                                  std::map<int, std::vector<hit> >& hits2wire,
```
it looks like the map is not modified by the function, can we pass it as a `const&` ?

gabriele-sirri commented on include/SANDGeoManager.h:
```diff
@@ -12,28 +15,45 @@
+
+namespace chamber
+{
+const char* const wire_regex_string =
```
why not `std::string` ? 

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#slstr1-use-stdstring-to-own-character-sequences

gabriele-sirri commented on include/SANDKalmanFilter.h:
```diff
@@ -0,0 +1,217 @@
+  int pdg_code;
+  int id;
+
+  TVector3 pos;
```
TVector3 is a legacy class.

https://root.cern.ch/doc/master/classTVector3.html

gabriele-sirri commented on include/SANDKalmanFilter.h:
```diff
@@ -0,0 +1,217 @@
+  TVector3 pos;
+  TVector3 mom;
+};
+
```
Instead of repeating `SANDKF... `for each type used in this file, this would be a great place to declare some namespaces to improve scoping and readability.
`sand::kf::Manager` `sand::kf::Track` `sand::kf::StateVector`




https://google.github.io/styleguide/cppguide.html#Namespaces

gabriele-sirri commented on include/SANDTrackerCell.h:
```diff
@@ -0,0 +1,109 @@
+
+class SANDTrackerCell
+{
+  SANDTrackerCellID _id;
```
several standards are used in this merge requests for class data members. Please consider standardizing the naming convention

https://google.github.io/styleguide/cppguide.html#Variable_Names

gabriele-sirri commented on include/SANDTrackerCell.h:
```diff
@@ -0,0 +1,109 @@
+  SANDTrackerCell() {};
+  SANDTrackerCell(const SANDTrackerCellID cID,
+                  const SANDWireInfo &l, 
+                  const double w,
```
`const double w`   -> `double w`

cheap to copy

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const

this comment applies to other functions defined here.

gabriele-sirri commented on include/SANDTrackerCell.h:
```diff
@@ -0,0 +1,109 @@
+  {
+    return _id;
+  }
+  void size(double &h, double &w) const
```
To return multiple “out” values, prefer returning a struct.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f21-to-return-multiple-out-values-prefer-returning-a-struct

gabriele-sirri commented on include/SANDGeoManager.h:
```diff
@@ -136,67 +171,127 @@ class SANDGeoManager : public TObject
+  long get_wire_id(long drift_plane_id, double z,
+                   double transverse_coord) const;
+  std::vector<SANDTrackerCellID> get_segment_ids(const TG4HitSegment& hseg) const;
+  TVector3 FindClosestDrift(TVector3 point, double epsilon) const;
```
TVector3 is a legacy class.

gabriele-sirri commented on include/SANDGeoManager.h:
```diff
@@ -136,67 +171,127 @@ class SANDGeoManager : public TObject
+  TVector3 FindClosestDrift(TVector3 point, double epsilon) const;
+  TVector3 SmearPoint(TVector3 point, double epsilon) const;
+  
+  const TVector2 pointInRotatedSystem(TVector2 v, double angle) const;
```
TVector2 is a legacy class as well.

https://root.cern.ch/doc/master/classTVector2.html

valerpia commented:
I will comment the last commit later today or tomorrow

valerpia commented:
Comments on https://github.com/DUNE/sandreco/pull/66/commits/9461b286f8e824277bc24982f81fae0d0b5b29de

**Completed:**
- added sand_reco::kf, sand_reco::tracker and sand_geometry::tracker namespaces where needed
- renamed namespaces to all lowercase
- renamed functions to camel case without the first letter capitalized
- renamed variables as requested (snake case with trailing underscores in classes and so on)
- renamed class to camel case
- applied most of the suggestions in the comments

**To Check:**
- const ref where needed
- cheap to copy where needed
- {} placement

**Not done/To be done but (IMO) not in this pull request:**
- replace the use of legacy classes (TVector3/2 and so on)
- apply all of the above to other sections of the code
 
Currently, I only implement these changes in the files/sections affected by this pull request. While this is necessary for the rest of the code too, I would propose doing it on a dedicated pull request to avoid going out of scope on this (already quite big) pull request

beamhalo reviewed (approved):
In absence of formal tests, and having verified that unrelated functions appear unaffected, I approve
